### PR TITLE
[BE] Deprecate `has_XYZ` attributes

### LINKED ISSUE #102484: Best practices clarification for determining accelerator availability e.g should one use  `torch.has_mps` or `torch.backends.mps.is_available()` and so on for cuda, etc
### 📚 The doc issue

What is the best way to check accelerator availability without knowing what system you are on / which version of torch is compiled?

Here is a very simplistic example

```python
accel = 'cpu'
if torch.backends.mps.is_available():
    accel = 'mps'
elif torch.cuda.is_available():
    accel = 'cuda'
```

Or should it be

```python
accel = 'cpu'
if torch.has_mps:
    accel = 'mps'
elif torch.has_cuda:
    accel = 'cuda'
```

### Suggest a potential alternative/fix

Clarify in the documentation somewhere. 

cc @svekars @carljparker @albanD @kulinseth @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by malfet**
Use [`__getattr__`](https://peps.python.org/pep-0562/) to raise warningwhen one tries to access `has_XYZ` methods and recommend appropriate `torch.backends.XYZ` methods

Make respective properties in `torch._C` private (by prefixing them with underscore), to exclude from `from torch._C import *`.

Added `warnings.simplefilter` to workaround Python-3.11 torch.compile lineinfo issue.

Fixes https://github.com/pytorch/pytorch/issues/102484

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -1708,5 +1708,18 @@ def registerOp(cls, op_key, full_schema, op_impl, dispatch_key):
...
 
 
+# Deprecated attributes
+_deprecated_attrs = {
+    "has_mps": torch.backends.mps.is_built,
+    "has_cuda": torch.backends.cuda.is_built,
+    "has_mkldnn": torch.backends.mkldnn.is_available,
+}
+def __getattr__(name):
+    v = _deprecated_attrs.get(name)
```
Comment (janeyx99) [inline torch/__init__.py:1718]:
```suggestion
    replacement = _deprecated_attrs.get(name)
```
Nit, but could we name v something more semantically close to what it represents?

**Review by janeyx99 — APPROVED** [review]
Looks good overall, though you'd need to convince mypy the _has* versions exist.

**Comment by malfet** [comment]
Surprisingly, `_dynamo` regressions are real, investigating...

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
LGTM assuming dynamo regressions can be sorted out.

**Comment by malfet** [comment]
For posterity: reproducer with running just 2 tests:
```
% TORCH_LOGS=bytecode PYTORCH_TEST_WITH_DYNAMO=1 python test_indexing.py  -v -k test_boolean_shape -k test_bool_indices_cpu
```

**Pull request closed without merging**